simple-ipc.c

```diff
@@ -182,4 +182,143 @@ int ipc_send_command(const char *path, const char *message, struct strbuf *answe
 		close(fd);
 	return ret < 0 ? -1 : 0;
 }
```

Comment:
nit: this commit could maybe use more text in the message body. Perhaps:

```
We previously implemented simple Inter-Process Communication (IPC) on
Windows using named pipes. Create an equivalent implementation for
Unix-based platforms using sockets.

(insert any subtleties about how this differs from the Windows implementation)

The tests in t0052-simple-ipc.sh will automatically start running on Unix
platforms with this change due to the response from "test-tool simple-ipc
SUPPORTS_SIMPLE_IPC" changing automatically with the change here in
simple-ipc.h.
```

Reply:
How about this:

    simple-ipc: implement a backend based on Unix-sockets

    We previously implemented simple Inter-Process Communication (IPC) on
    Windows using named pipes. Create an equivalent implementation for
    Unix-based platforms using Unix sockets.

    Note that there are a variety of differences between Win32 Named Pipes
    and Unix sockets, for example:

    - a Win32 Named Pipe vanishes once the listener handle is closed. Unix
      sockets do not just disappear, even if the process that created them
      exited.

    - Win32 Named Pipes exist in a name space parallel to the actual file
      system. In contrast, Unix sockets live in the same name space, i.e. a
      Unix socket cannot have the same path as an existing file and vice
      versa.

    - Win32 Named Pipes have two communication modes: "message" and "byte".
      The former allows senders to define the size of messages to be sent;
      the receiving end will not receive partial messages. Unix sockets only
      operate in the mode equivalent to "byte".

    - When a listener accepts a client via a Win32 Named Pipe, subsequent
      communication will use the same handle to the same pipe, i.e. it can
      become a bit tricky to keep the communication to several concurrent
      clients apart from one another. When a listener on a Unix socket
      accepts a new client, it gets handed a new file descriptor specific to
      that client.

    The tests in `t0052-simple-ipc.sh` will automatically start running on
    Unix platforms with this change due to the response from `test-tool
    simple-ipc SUPPORTS_SIMPLE_IPC` changing automatically with the change
    here in `simple-ipc.h`.

Reply:
WRT:

```
    - When a listener accepts a client via a Win32 Named Pipe, subsequent
       communication will use the same handle to the same pipe, i.e. it can
       become a bit tricky to keep the communication to several concurrent
       clients apart from one another. When a listener on a Unix socket
       accepts a new client, it gets handed a new file descriptor specific to
       that client.
```

Windows Named Pipes actually behave a little differently than that.
The server creates the initial server-pipe (which reserves the pathname
in the Named Pipe File System) and allows it to receive the first client
connection.  That part is true.  But when either side closes that connection,
the pipe is dead.  The pathname is removed from the NPFS when the other
side closes their handle.

For the server to respond to multiple clients (either in series or currently),
it needs to open more than one server pipe handles on the same pathname.
(The server should set the ACL on the initial server-pipe handle, because the
ACL can't be changed once there are multiple handles.)

So yes, the server needs to juggle a list/queue of open server handles.  The next
incoming client connection will connect (randomly to) one of the open
server handles.  The server needs to pull that from the "ready list/queue" and
dispatch it to a worker thread for the duration of the conversation/protocol.
The server needs to open a new server handle on the existing pathname
to replace the consumed one in the queue.

Most of the example code on the net show a server opening a single server
handle, waiting for a connection, dispatching it and then opening a new
single server handle to replace it (and before the worker thread can close
the connected handle).  There are 2 races in that model: (1) the worker hangs
up too quickly and (2) the named pipe is non-responsive during that the interval
(after the connection is accepted and until the new server handle is opened).

(Creating a named-pipe client handle does secretly have a timeout in it, but
it is still possible for the client to timeout and fail to get a client handle (when
you have a high arrival rate (ask me how I know...)))

By letting the server pre-create multiple server handles (mostly) prevents
clients from starving, and implicitly prevents the other race.  Conceptually,
this server handle list/queue acts like the listen(n) in the socket code.

Reply:
@jeffhostetler can I ask you to implement this? You are the domain expert about this, and I don't trust myself to get it right...

Reply:
Okay.